### PR TITLE
Rate limit all unauthenticated HTTP endpoints

### LINKED ISSUE #4330: Rate-limit proxy web API endpoints
### Feature Request

Add a rate-limiter on all web API endpoints, with default limit for all and overrides for specific endpoints.

### Motivation

Some web API endpoint are DoS vectors - they do a large amount of work, which slows down more with contention (repeated/parallel requests). An example is `webapi/ping`, which can consume a lot of FDs.

### Who's it for?
OSS User, Pro, Enterprise

### PR DESCRIPTION
**Pull request by jentfoo**
This PR is an extension to what was done in PR #172. This PR is designed to fix https://github.com/gravitational/teleport/issues/4330 and also fix https://github.com/gravitational/teleport-private/issues/403.

Rather than audit endpoints and choose what endpoints should be rate limited, this commit proposes that for safety and reduced cognitive load, all unauthenticated endpoints become rate limited.

The primary concern in this type of change would be if our rate limit becomes too aggressive for general use.  There are two considered strategies to make sure this does not become impacting:
1. Adjust the rate limiter so the rate limit becomes endpoint specific.  This would avoid the need to consider how activity on one endpoint effects another.
2. Accept that rate limit interactions are possible and instead ensure rate limits are high enough to avoid this concern.

This commit chooses option 2.  While `opt 1` has advantages, particularly as endpoints and new use cases are added.  Option 2 provides the strictest and safest rate limits.

Our rate limits were configured to:
period: 1 min
avg rate: 10
burst rate: 20

In order to build a safety buffer with option #2 those allowed rates were doubled.

### CODE REVIEW

File: lib/defaults/defaults.go
Comment (zmb3) [inline lib/defaults/defaults.go:371]: These values are no longer specific to passwordless interactions, are they?

Comment (jentfoo) [reply lib/defaults/defaults.go:371]: AFAIK they are only used in passwordless interactions, but there may be something I am missing.

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:524]: I would be a little bit concerned about limiting the ping endpoint, I think it may be used as part of the node joining or initial connection process. (we should confirm whether this is the case or not)

For example, suppose a customer has thousands of nodes in some private network with SNAT. It will appear as if all of these nodes are connecting from the same source IP and could trip the rate limiter and prevent nodes from joining or reconnecting after an upgrade.

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:622]: You forgot the auth export endpoints on line 618-619

**Comment by zmb3** [comment]
A change like this should be reserved for a major release and run through the full test plan including load testing and performance tests.

This process will begin for Teleport 13 as early as Monday, which means this might have to wait until Teleport 14.

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go:524]: https://github.com/gravitational/teleport/issues/4330 specifically calls out the file descriptors that are opened from ping.  I understand the concern though.

In such a condition, would the reconnecting nodes retry until they are able to eventually get a rate allowed request?

Do you think any rate limit could be designed around ping?

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go:622]: Good spot, addressed in 454925b67e5d2364622ded3cd569f38ddd94ba8d

**Comment by jentfoo** [comment]
I agree with the test plan @zmb3.  I wont merge this to `master` until 13.0.0 has been released

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go:524]: @zmb3 As an alternative solution, could the nodes start to send their `ping` request authenticated?  If so we could still make this endpoint "optionally unauthenticated", but let the rate limiter ignore requests which provide valid authentication.

File: lib/defaults/defaults.go
Comment (Joerger) [reply lib/defaults/defaults.go:371]: Looking at the code, I think Zac is right, so the `Passwordless` part should be removed.

**Review by Joerger — COMMENTED** [review]
Should we also add rate limiters to the enterprise only endpoints? For example, it looks like we need one here - https://github.com/gravitational/teleport.e/blob/53ef3ae994b5968b980fdff9704dae9baeb56ba4/lib/web/plugin.go#L163.

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go]: I added the ability to only rate limit unauthenticated requests in this commit: https://github.com/gravitational/teleport/pull/24623/commits/289b6114f5e6acd4749be8ef0bd046881ec077dd

As mentioned in the message luckily `ping` was already providing authentication, so this logic worked pretty straight forward out of the box.  I added this to other endpoints which _might_ be good candidates to consider, although I did not verify they were providing auth like `ping` was.

Please provide feedback on this change, thank you!

**Comment by jentfoo** [comment]
> Should we also add rate limiters to the enterprise only endpoints?

I will submit a PR to update to use the new `WithUnauthenticatedLimiter` once we merge this PR.  Thank you for spotting @Joerger!

File: lib/web/apiserver.go
Comment (Joerger) [inline lib/web/apiserver.go:3656]: Is there a reason you removed the `trace.Wrap`?
```suggestion
			return nil, trace.Wrap(err)
```

File: lib/web/apiserver.go
Comment (Joerger) [inline lib/web/apiserver.go:3637]: nit:
```suggestion
		if _, _, err := h.authenticateRequestWithCluster(w, r, p); err != nil {
			// retry with user auth
			if _, err = h.AuthenticateRequest(w, r, true); err != nil {
				// no auth passed, limit request
				return h.WithLimiterHandlerFunc(fn)(w, r, p)
			}
		}
```

**Review by Joerger — APPROVED** [review]

**Comment by jentfoo** [comment]
@zmb3 Would you mind reviewing this PR?

File: lib/auth/middleware.go
Comment (zmb3) [inline lib/auth/middleware.go:386]: Is there any reason not to also use these defaults for the account recovery RPCs above?

As a sidenote, it looks like the changes in this PR are exclusive to HTTP endpoints, and not RPCs. Might be worth clarifying in the PR description / commit message.

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:524]: I'm still a little nervous about ping. `tsh login` hits this endpoint, so with the current settings, no more than 40 users can log in (from the same IP) during a 1 minute window. 

There are several ways (SNAT, VPN gateway, etc) that it could appear that all users are connecting from the same IP. 

What do you think about either using more generous limits for this endpoint, or starting with some metrics/logging and otherwise leaving it unlimited so that we can evaluate the impact of such a change?

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:3667]: Nit: Technically, this is a type assertion, not a cast.

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:3910]: Why did you decide to implement manual limiting here inside the handler as opposed to using a middleware like you did for all the other endpoints?

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:3641]:
```suggestion
// WithLimiter adds IP-based rate limiting to fn.
// Limits are applied to all requests, authenticated or not.
```

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:3631]:
```suggestion
			if _, err = h.AuthenticateRequest(w, r, true /* check token */) ; err != nil {
```

File: lib/auth/middleware.go
Comment (jentfoo) [reply lib/auth/middleware.go:386]: > Is there any reason not to also use these defaults for the account recovery RPCs above?

I don't see any reason to increase those rates, do you?

> As a sidenote, it looks like the changes in this PR are exclusive to HTTP endpoints

Good point, I will update the descriptions

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go:524]: > What do you think about either using more generous limits for this endpoint, or starting with some metrics/logging and otherwise leaving it unlimited so that we can evaluate the impact of such a change?

I believe this is one of the more important endpoints to rate limit.  What rate limit would make you comfortable?  We could potentially keep the average at 20, but increase the burst to 100 or 120.

The goal in having this land right after v13 is cut is so that we can have as long as possible period to test and witness any unexpected interactions from this change.

Let me know your thoughts, thank you!

Comment (zmb3) [reply lib/web/apiserver.go:524]: I don't know what limits would make me more comfortable, because I don't have any data to inform me as to what's reasonable and expected and what's abnormal.

If we put a prometheus metric in for requests/s to this endpoint (grouped by remote IP) we wouldn't have to collect data very long to get a good idea of what's normal.

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go]: Updated the comment, good spot

File: lib/web/apiserver.go
Comment (jentfoo) [reply lib/web/apiserver.go]: @zmb3 and @Joerger Can you please review the latest commit: [3d31f86](https://github.com/gravitational/teleport/pull/24623/commits/3d31f86f24f890b6cf21d0c563b20362ea62c1da)

There I introduce a `high` rate limiter for these select endpoints which are only CPU bound.  After discussions with @zmb3 I believe this to be the safest balance, but please give me your candid thoughts.  Thank you!

**Review by zmb3 — APPROVED** [review]

**Pull request merged**
